org-brain.el

```diff
@@ -62,24 +64,94 @@ This will be used by `org-brain-new-child'."
   :group 'org-brain
   :type '(string))
 
+;;; Utils
+(defun flatten (obj)
+  (do* ((result (list obj))
+        (node result))
+       ((null node) (delete nil result))
+    (cond ((consp (car node))
+           (when (cdar node) (push (cdar node) (cdr node)))
+           (setf (car node) (caar node)))
+          (t (setf node (cdr node))))))
+
+;;; Logging
+(defcustom org-brain-log t
+  "Set to nil to not write to *Messages* buffer."
+  :group 'org-brain
+  :type 'boolean)
+
+(defun org-brain-log (msg)
+  (when org-brain-log
+    (setq inhibit-message t)
+    (message msg)
+    (setq inhibit-message nil)))
+
+;;; Caches
+(setq org-brain-files-cache nil)
+(setq org-brain-parents-cache nil)
+(setq org-brain-children-cache nil)
+(setq org-brain-pins-cache nil)
+
+(defun org-brain-invalidate-all-caches ()
+  "This is a convenience function for those who (occasionally)
+  edit children, parents, i.e., entries in the Brainchildren
+  node, or pins, outside the org-brain-visualize interface. In
+  that case, you have to call this function manually. It is not
+  needed if children, parents, and pins are added using the
+  org-brain-visualize interface/mode."
+  (interactive)
+  (setq org-brain-files-cache nil)
+  (setq org-brain-parents-cache nil)
+  (setq org-brain-children-cache nil)
+  (setq org-brain-pins-cache nil))
+
+(defun org-brain-invalidate-files-cache ()
+  (org-brain-log "Invalidating org-brain file cache...")
+  (setq org-brain-files-cache nil))
+
+(defun org-brain-invalidate-parent-cache-entry (entry)
+  (org-brain-log (format "Invalidating org-brain parent cache entry: %s ..." entry))
+  (setq org-brain-parents-cache
+        (remove* entry org-brain-parents-cache :test #'equal :key #'car)))
+
+(defun org-brain-invalidate-child-cache-entry (entry)
+  (org-brain-log (format "Invalidating org-brain child cache entry: %s ..." entry))
+  (setq org-brain-children-cache
+        (remove* entry org-brain-children-cache :test #'equal :key #'car)))
+
+(defun org-brain-invalidate-pins-cache ()
+  (org-brain-log "Invalidating org-brain pin cache...")
+  (setq org-brain-pins-cache nil))
+
 (defun org-brain-files (&optional relative)
   "Get all org files (recursively) in `org-brain-path'.
 If RELATIVE is t, then return relative paths and remove org extension."
-  (make-directory org-brain-path t)
-  (let ((files (directory-files-recursively
-                org-brain-path (format "\\.%s$" org-brain-files-extension))))
-    (if relative
-        (mapcar #'org-brain-path-entry-name files)
-      files)))
+  (if org-brain-files-cache
+      (if relative
+          (mapcar #'org-brain-path-entry-name org-brain-files-cache)
+        org-brain-files-cache)
+    (progn
```

Comment:
This `progn` could be `unless org-brain-files-cache` in beginning of function, and after updating the cache run the `if relative` sexp.